### ISSUE TITLE
Running a new thread in Fiber.schedule block raises FiberError

### ISSUE THREAD

**Issue by kudojp** [kudojp/async_std_scheduler#31]
## Describe the bug

Snippet below fails with FiberError.

```rb
Fiber.set_scheduler AsyncScheduler::Scheduler.new
Fiber.schedule do
    Thread.new{}.join
end
```

## To Reproduce

Open a console by `bin/console`, and then run below.

```rb
[1] pry(main)> Fiber.set_scheduler AsyncScheduler::Scheduler.new
=> #<AsyncScheduler::Scheduler:0x0000000102f239d0 @blockings=#<Set: {}>, @input_waitings={}, @output_waitings={}, @waitings={}>
[2] pry(main)> Fiber.schedule do
[2] pry(main)*   Thread.new{}.join
[2] pry(main)* end  
#<Thread:0x00000001064c72a8 (pry):3 run> terminated with exception (report_on_exception is true):
/Users/daiki-kudo/repos/async_scheduler/lib/async_scheduler/scheduler.rb:49:in `resume': attempt to resume a resumed fiber (double resume) (FiberError)
        from /Users/daiki-kudo/repos/async_scheduler/lib/async_scheduler/scheduler.rb:49:in `unblock'
FiberError: attempt to resume a resumed fiber (double resume)
from /Users/daiki-kudo/repos/async_scheduler/lib/async_scheduler/scheduler.rb:49:in `resume'
```

## Expected behavior

It does not raise an error.

**Comment by kudojp** [comment]
> FiberError: attempt to resume a resumed fiber (double resume)

The same error message can be found in the final snippet in this page:
https://docs.ruby-lang.org/en/master/Fiber.html


Shortly saying, FiberError (double resume) is raised when a Fiber(X) is executed in another fiber(Y) which is resumed by Fiber X. That is, that Fiber X is an ancestor of Fiber Y


```rb
f2 = nil

f1 = Fiber.new do
  f2.resume
end

f2 = Fiber.new do
  f1.resume
end

f2.resume

# => test2.rb:4:in `resume': attempt to resume a resumed fiber (double resume) (FiberError)
#           from test2.rb:4:in `block in <main>'
```

Note that when X = Y, it raises FiberError but its error message is not explain it as the double resume.

```rb
f1 = Fiber.new do
  f1.resume
end

f1.resume
# => test2.rb:2:in `resume': attempt to resume the current fiber (FiberError)
#           from test2.rb:2:in `block in <main>'

**Comment by kudojp** [comment]
# The root cause

I think could understand what is going on internally.

## TL; DR

FiberScheduler registered in the parent thread is somehow inherited to the child thread 🟦. 
A fiber task created in the parent thread (fiber X) resumes a root fiber in the child thread (fiber Y). In fiber Y, the scheduler somehow shared with the parent thread is working. It resumes fiber X, which raises FiberError (double resume).

## Reproduction with debugging logs

1. Checkout to 2b89949c0130f671b90c5238675cfcffbc4d648f
2. Run $ bundle exec ruby [test.rb](https://github.com/kudojp/async_scheduler/blob/2b89949c0130f671b90c5238675cfcffbc4d648f/test.rb)

This stdouts as blow. This is formatted before being pasted here.

```
#<Thread:0x000000010f3d3de8 test.rb:10 run> terminated with exception (report_on_exception is true):
/Users/daiki-kudo/repos/async_scheduler/lib/async_scheduler/scheduler.rb:59:in `resume': attempt to resume a resumed fiber (double resume) (FiberError)
        from /Users/daiki-kudo/repos/async_scheduler/lib/async_scheduler/scheduler.rb:59:in `unblock'

### thread_ids ####
{
  :parent_thread_id=>1400,
  :child_thread_id=>1480
}

### scheduler_ids ####
{
  :parent_scheduler_id=>1420,
  :child_scheduler_id=>8                 # memo: 8 is an object_id of nil
}

### All points where tasks are resumed ####
From now, resume a fiber (id=1440) at /Users/daiki-kudo/repos/async_scheduler/lib/async_scheduler/scheduler.rb#32 
    inside of Scheduler (id=1420) in Thread (id=1400).
    Currently in the fiber (id=1460). Callers are: ["test.rb:9:in `schedule'", "test.rb:9:in `<main>'"]
From now, resume a fiber (id=1440) at /Users/daiki-kudo/repos/async_scheduler/lib/async_scheduler/scheduler.rb#58 
    inside of Scheduler (id=1420) in Thread (id=1480).
    Currently in the fiber (id=1500). Callers are: []
```

## My understandings 

1. A parent thread (id=1400) starts. This thread holds a FiberScheduler (id=1420).
4. In the parent thread, Fiber instance (id=1440) is created, and executed until it is blocked at created thread's #join.
5. In the parent thread, all the lines after Fiber.schedule are executed. Obviously there is no line in this case.
6. In the parent thread, the parent thread's FiberScheduler#close is called and the event loop starts. The pending Fiber task (id=1440) is resumed again. Note that the task has not been removed yet from @blockings in the FiberScheduler.
7. In the task (id=1440), it calls Thread#join method. This forces the child thread to be finished, so the execution of the child thread resumes. The root Fiber in this child thread is probably id=1500, and this Fiber (id=1500) is under the Fiber (id=1440) 🟥.
8. The child thread (id=1480) is executed until the end. This time there's no line.
9. The child thread seems to share the FiberScheduler (id=1420) with the parent thread 🟦 and at this moment, the child thread's FiberScheduler#close seems to be called. This executes the task (id=1440) which is still held in @blockings of the shared FiberScheduler. As the fact written in 🟥 tells, it tries to resume its parent Fiber task, so it throws FiberError (double resume).


## Notes

My understanding is not necessarily correct.

> FiberScheduler registered in the parent thread is somehow inherited to the child thread 🟦. 

If this is correct, the stdout of $ bundle exec ruby [test.rb](https://github.com/kudojp/async_scheduler/blob/2b89949c0130f671b90c5238675cfcffbc4d648f/test.rb) should NOT be like:

```
### scheduler_ids ####
{
  :parent_scheduler_id=>1420,
  :child_scheduler_id=>8                 # memo: 8 is an object_id of nil
}
```

but it should be:

```
### scheduler_ids ####
{
  :parent_scheduler_id=>1420,
  :child_scheduler_id=>1420                 
}
```

**Comment by kudojp** [comment]
# How to fix this issue

Shortly saying, the current version (e54d9aabf1176a9433a7a94534b057250ceb771b) of AsyncScheduler is not thread safe.
- For the short term solution, make the scheduler raise an error when it is used in multithread programming. 
   - 👉  Patched in https://github.com/kudojp/async_scheduler/pull/29
- For the long term solution, make the scheduler to be thread safe.
